File: nim/file2bin/file2bin.nim
#
# Extract binary portion from file to file
# Copyright (c) 2021 Dmitry Stefankov
#
import os
import strutils
#import parseutils
import streams

echo "program name: ", getAppFilename()
#echo "Arguments:"
#for arg in commandLineParams():
#  echo arg
#echo paramCount(), " ", paramStr(1)
let args_count =  paramCount()
if args_count < 4:
  echo "Extract binary portion from file, version 1.0"
  echo "Copyright (C) 2021 Dmitry Stefankov. All Rights Reserved."
  echo "Usage: file2bin.exe infile outfile offset size"
  quit(1)

let infile = paramStr(1)
let outfile = paramStr(2)
echo "infile: ", infile
echo "outfile: ", outfile
let offset = paramStr(3).parseInt()
let length = paramStr(4).parseInt()
echo "offset: ", offset
echo "length: ", length

let filesize = getFileSize infile
echo "filesize: ", filesize

var mybuf = alloc(length)

var fs = newFileStream(infile, fmRead)
fs.setPosition(offset)
var readlen = readData(fs, mybuf, length)
echo "readlen: ",readlen
fs.close()

var newfs = newFileStream(outfile, fmWrite)
writeData(newfs, mybuf, length)
newfs.close()

File: R/file2bin/file2bin.r
# Rscript[.exe] file2bin.r test.in test.out 35 67

Sys.setenv(LANG='C')

main <- function() {
  args <- commandArgs(trailingOnly = TRUE)
  in_filename <- args[1]
  out_filename <- args[2]
  in_offset <- args[3]
  in_size <- args[4]
  print(in_filename)
  print(out_filename)
  print(in_offset)
  print(in_size)
  cx <- file(in_filename, "rb")
  seek(cx, in_offset)
  d <- readBin(cx,"raw",in_size,size=1)
  close(cx)
  con <- file(out_filename, "wb")
  writeBin(d, con, size = 1)
  close(con)
}

if (length(commandArgs(trailingOnly = TRUE)) == 0) {
  print("Extract binary portion from file, version 1.0")
  print("Copyright (C) 2024 Dmitry Stefankov. All Rights Reserved.");
  print("Usage: file2bin infile outfile in_offset in_size");
  print("infile    - input filename");
  print("outfile   - output filename");
  print("in_offset - byte offset on input stream (default=0)");
  print("in_size   - copy N bytes from input stream (default=0)");
  quit(status=1)
}

main()



